### PR TITLE
implement type_boolean macro

### PR DESCRIPTION
**Pull request by jpmmcneill**
[WIP]

resolves (partially) #5739

### Description

dbt-core component required for #5739.

_Other PRs_
- dbt-redshift: https://github.com/dbt-labs/dbt-redshift/pull/190
- dbt-bigquery: https://github.com/dbt-labs/dbt-bigquery/pull/313
- dbt-snowflake: https://github.com/dbt-labs/dbt-snowflake/pull/268
- dbt-spark: https://github.com/dbt-labs/dbt-spark/pull/471

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: core/dbt/include/global_project/macros/utils/data_types.sql
Comment (jpmmcneill) [inline core/dbt/include/global_project/macros/utils/data_types.sql:62]: every other comment had this many dashes 🙃

**Comment by dbeatty10** [comment]
@jpmmcneill this is looking good to me! 🤩 

Here's the next steps for this PR:
1. Run `changie new` per [these](https://github.com/dbt-labs/dbt-core/blob/e00eb9aa3a173ecf40f0e0190fb496a64d3ca876/CONTRIBUTING.md#adding-a-changelog-entry) instructions and commit and push the resulting file.
1. Open up pull requests for dbt-redshift, dbt-snowflake, and dbt-spark that add [your new test](https://github.com/dbt-labs/dbt-bigquery/pull/313/files#diff-4937071b580318527a09aeee7aaf4abf2f3be690e0263ee415d45cee90ac12afR8-R38) to their test suites.
    - Unfortunately, new tests aren't inherited and executed automatically -- they need to be added explicitly.

I'll add some further instructions on your [PR for dbt-bigquery](https://github.com/dbt-labs/dbt-bigquery/pull/313) on how to point at this PR branch for the purposes of automated CI tests in GitHub. (We'll want the adapters to be using your PR branch rather than the `main` branch.)

Once all of the PRs are passing the tests in GitHub Actions, we will merge dbt-core first. Then we will run CI once more for each adapter (pointing back at `main`) before merging them.

File: tests/adapter/dbt/tests/adapter/utils/data_types/test_type_boolean.py
Comment (dbeatty10) [inline tests/adapter/dbt/tests/adapter/utils/data_types/test_type_boolean.py:10]: Probably want to cover both boolean options at the very least.
```suggestion
seeds__expected_csv = """boolean_col
True
False
""".lstrip()

models__actual_sql = """
select cast('True' as {{ type_boolean() }}) as boolean_col
union all
select cast('False' as {{ type_boolean() }}) as boolean_col
"""
```

### Even better
But test cases covering the truth tables for [conjunction, disjunction, and negation](https://en.wikipedia.org/wiki/Boolean_algebra#Basic_operations) would be even better. That way, we are verifying that everything is _acting_ like booleans.

image image image

Something like the following **untested** code:
```
seeds__boolean_permutations_csv = """
x,y
False,False
True,False
False,True
True,True
""".lstrip()

seeds__expected_csv = """
x,y,conjunction,disjunction,negation_x
False,False,False,False,True
True,False,False,True,False
False,True,False,True,True
True,True,True,True,False
""".lstrip()

models__actual_sql = """
select
    x,
    y,
    x and y as conjunction,
    x or y as disjunction,
    not x as negation_x
from {{ ref("boolean_permutations" }}
```

### Best? 🤷 

Even though `BOOLEAN` is a data type described in the SQL standard, some databases don't have it (looking at you, SQL Server!). 

If we are feeling extra magnanimous, we could change all the True/False values in the seeds to be `1`/`0` instead.

I'm hoping it wouldn't be necessary, but we could update the `models__actual_sql` definition so that x/y values are replaced with the following instead:
- `cast(x as {{ type_boolean() }})`
- `cast(y as {{ type_boolean() }})`

Comment (dbeatty10) [reply tests/adapter/dbt/tests/adapter/utils/data_types/test_type_boolean.py:10]: Update:

The "Even better" example I gave above didn't actual test the `type_boolean()` macro at all! 😅 

Something like this should fix that situation:
```
models__actual_sql = """
select
    cast(x as {{ type_boolean() }}) as x_bool,
    cast(y as {{ type_boolean() }}) as y_bool,
    cast(x as {{ type_boolean() }}) and cast(y as {{ type_boolean() }}) as conjunction,
    cast(x as {{ type_boolean() }}) or cast(y as {{ type_boolean() }}) as disjunction,
    not cast(x as {{ type_boolean() }}) as negation_x
from {{ ref("boolean_permutations" }}
```

Comment (jpmmcneill) [reply tests/adapter/dbt/tests/adapter/utils/data_types/test_type_boolean.py:10]: Cheers @dbeatty10. I'll take a second pass at this later today :)

**Review by dbeatty10 — APPROVED** [review]
@jpmmcneill I just realized that your simple pytest for `type_boolean()` was exactly in line with the rest of the `type_x` macros.

So I'm approving as-is.

After this PR is merged, then the next steps will be to restore the original dev-requirements.txt files in each of your adapter PRs.

**Comment by jpmmcneill** [comment]
> @jpmmcneill I just realized that your simple pytest for `type_boolean()` was exactly in line with the rest of the `type_x` macros.
> 
> So I'm approving as-is.
> 
> After this PR is merged, then the next steps will be to restore the original dev-requirements.txt files in each of your adapter PRs.

Hey @dbeatty10 - thanks. I personally completely agree with the sentiment around the current level of testing :).

Do you agree with me that an issue that basically scopes "improve the current test coverage for data types" would be welcome?

**Comment by dbeatty10** [comment]
> Do you agree with me that an issue that basically scopes "improve the current test coverage for data types" would be welcome?

Formally logging where expectations differs from reality is an important mechanism for us. A new issue describing the current level of testing and how that compares to what your expectations were as a contributor ([here](https://github.com/dbt-labs/dbt-core/pull/5875)) and dbt package maintainer ([here](https://github.com/jpmmcneill/dbt-graph-theory)) would be great!

From there, someone from dbt Labs (maybe me!) will triage the issue submission and give feedback, determine urgency, compare to current roadmap and capacity, etc.

**Comment by jpmmcneill** [comment]
Brill, will do. Thanks @dbeatty10 🐐

**Pull request merged**
